secrets/secrets.go

```diff
@@ -298,3 +312,56 @@ func NewSecrets(r io.Reader) (*Secrets, error) {
 	}
 	return secrets, nil
 }
+
+func csiPathParser(path string, inputSecrets Document) (secretsDocument Document, err error) {
+	secretsDocument = Document{
+		Secrets: make(map[string]GenericSecret),
+	}
+	file, err := os.Open(path)
+	if err != nil {
+		return secretsDocument, err
+	}
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return secretsDocument, err
+	}
+	if fileInfo.IsDir() {
+		files, err := os.ReadDir(path)
+		if err != nil {
+			return secretsDocument, err
+		}
+		for _, f := range files {
+			// recursive step
+			parsedSecrets, err := csiPathParser(filepath.Join(path, f.Name()), inputSecrets)
+			if err != nil {
+				return secretsDocument, err
+			}
+
+			for k, v := range parsedSecrets.Secrets {
+				inputSecrets.Secrets[k] = v
+			}
+		}
+		secretsDocument = inputSecrets
+		return secretsDocument, err
+	}
+	// parse file
+	var secretFile CSIFile
+	err = json.NewDecoder(file).Decode(&secretFile)
+	if err != nil {
+		return secretsDocument, err
+	}
+	secretsDocument.Secrets[path] = secretFile.Secret
+
+	return secretsDocument, err
+}
+
+func checkForVaultCSI(secretsDocument Document, err error) (Document, error) {
+	if e, ok := err.(*fs.PathError); ok {
+		// check if the path is a directory, then assume
+		// the secret provider is Vault CSI
+		if strings.Contains(e.Error(), "is a directory") {
```

Comment:
what you want here is `errors.Is(err, syscall.EISDIR)`, example: https://go.dev/play/p/VkiHFhVeQhI

but I still think that using the error to pass the path to the directory is a very anti-pattern way to do that and we should have made the distinction long before getting the error.

this is basically the last resort if we do not want to change the parser api in anyway (for a directory we still just pass in an `io.Reader` to the parser). I think we should make a different api for directory parser instead. but if that turns out to be infeasible then this is... probably fine.

Reply:
+1 on using `errors.Is` API. Forcing an error to handle this does not sound great. If we look here the path is being passed in before delegating it to the parser method (which is not exported): https://github.com/reddit/baseplate.go/blob/master/secrets/store.go#L48-L62

Couldn't we extend the unexported `parse` method here and add a new `newSecrets` constructor which also accepts a path?